internal/server/logbuffer/merge.go

```diff
@@ -0,0 +1,146 @@
+package logbuffer
+
+import (
+	"io"
+	"time"
+)
+
+// Merger can combine multiple log streams into one stream. It presumes
+// each stream emits entries in time order, and then weaves the entries
+// together to create a total time ordered stream.
+type Merger struct {
+	readers []MergeReader
+	heads   []TimedEntry
+}
+
+// TimedEntry is the interface each input returns entries in
+type TimedEntry interface {
+	// Time is the associated time of the value. This is used to sort
+	// the entry against other entries from other inputs.
+	Time() time.Time
+
+	// Value returns the log entry itself. This allows time addition
+	// wrappers to return their original value.
+	Value() interface{}
+}
+
+// MergeReader is value that returns TimedEntry's for LogMerge
+// to weave together.
+type MergeReader interface {
+	Next() (TimedEntry, error)
+}
+
+// NewMerger creates a new LogMerge, with the stream generated
+// from the given inputs.
+func NewMerger(readers ...MergeReader) *Merger {
+	var lm Merger
+	lm.readers = readers
+
+	lm.heads = lm.makeHeads()
+
+	return &lm
```

Comment:
nit: could we simplify this to 

      return &Merger{
               readers: readers,
               heads: make([]TimedEntry, 0, len(readers)),
     }

Reply:
Could!